### PR TITLE
Skip ObjectiveC/C++ tests if ObjC compiler is not available

### PR DESCRIPTION
**Pull request by tp-m**

### CODE REVIEW

File: run_project_tests.py [hunk 1]
```diff
@@ -356,6 +356,13 @@ def have_d_compiler():
         return True
     return False
 
+def have_objc_compiler():
+    if shutil.which("objc") and shutil.which("objcpp"):
+        return True
```
Comment (jpakkane) [inline run_project_tests.py]: This does not work. On debianlike machines there is no executable called `objc`. You just need to know that `gcc` masquerades as objc.

Comment (tp-m) [inline run_project_tests.py]: Any idea to check that then?

I have normal gcc installed of course but not the gobjc* packages, and the tests fail for me currently.

Comment (jpakkane) [inline run_project_tests.py]: The only reliable way is probably to try to compile an ObjC file.

Comment (nirbheek) [inline run_project_tests.py]: The best way is to do what we do in the unit tests: [`Environment.detect_objc_compiler()`](https://github.com/mesonbuild/meson/blob/8b73d807924dde250f1c4340885ef46a22b82ba7/run_unittests.py#L555).

File: run_project_tests.py [hunk 2]
```diff
@@ -405,6 +405,33 @@ def have_d_compiler():
...
+            return False
+        try:
+            objc_comp.sanity_check(env.get_scratch_dir(), env)
+        except:
+            return False
+        objcpp_comp = env.detect_objc_compiler(False)
+        if not objcpp_comp:
+            return False
+        try:
+            objcppcomp.sanity_check(env.get_scratch_dir(), env)
```
Comment (ignatenkobrain) [inline run_project_tests.py]: __[flake8]__
* [F821] undefined name 'objcppcomp'

**Comment by tp-m** [comment]
New version, but I'm not going to lie, this is just me trying random things. Works for me of course.

File: run_project_tests.py [hunk 3]
```diff
@@ -405,6 +405,33 @@ def have_d_compiler():
         return True
     return False
 
+def get_fake_options():
```
Comment (nirbheek) [inline run_project_tests.py]: You can import this from `run_tests.py` now.

File: run_project_tests.py [hunk 4]
```diff
@@ -427,7 +454,7 @@ def detect_tests_to_run():
     all_tests.append(('vala', gather_tests('test cases/vala'), False if backend is Backend.ninja and shutil.which('valac') else True))
     all_tests.append(('rust', gather_tests('test cases/rust'), False if backend is Backend.ninja and shutil.which('rustc') else True))
     all_tests.append(('d', gather_tests('test cases/d'), False if backend is Backend.ninja and have_d_compiler() else True))
-    all_tests.append(('objective c', gather_tests('test cases/objc'), False if backend in (Backend.ninja, Backend.xcode) and not mesonlib.is_windows() else True))
+    all_tests.append(('objective c', gather_tests('test cases/objc'), False if backend in (Backend.ninja, Backend.xcode) and have_objc_compiler() and not mesonlib.is_windows() else True))
```
Comment (nirbheek) [inline run_project_tests.py]: Don't need to check `is_windows` anymore.

File: run_project_tests.py (hunk 2 above)
Comment (nirbheek) [inline run_project_tests.py]: Typo...

**Comment by tp-m** [comment]
Seems to me (and appveyor) like we still need the is_windows test? Fixing the objective-c tests on windows is not really directly related to this PR, is it? Should be done separately if someone cares IMHO.

**Comment by jpakkane** [comment]
This message is posted to every outstanding merge request.

We have transitioned from Github wiki to our new documentation system that generates all documentation directly from the Git repo. This means that from now on every merge request must come with corresponding documentation changes.

If your MR requires documentation changes, do the necessary changes to files in `docs/markdown`.

If your MR adds a new feature, add a description to the release notes in `docs/markdown/Release-notes-for-0.41.0.md`.

If this commit is just a bugfix with no functional changes, you don't need to do anything.

Thank you

**Comment by jpakkane** [comment]
:crossed_swords: conflicts.

**Comment by nirbheek** [comment]
> Seems to me (and appveyor) like we still need the is_windows test? Fixing the objective-c tests on windows is not really directly related to this PR, is it? Should be done separately if someone cares IMHO.

Fair enough. We should get this in then, once conflicts are fixed.

**Pull request merged**